### PR TITLE
Make brain tests pass.

### PR DESCRIPTION
**Pull request by andreas-kupries**
## Description

See referenced issues.
See also the subordinate PR https://github.com/SUSE/brain-tests-release/pull/6

## Motivation and Context

  - https://jira.suse.com/browse/CAP-526
  - https://github.com/SUSE/kubecf/issues/29

## How Has This Been Tested?

Currently manual rebuild of the brain test release and going through the failing tests.
Details in the CAP ticket.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code has security implications.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by bikramnehra** [comment]
There is a new release out: https://github.com/SUSE/brain-tests-release/releases/tag/v0.0.8
_Mostly dealing with golang bump._

File: deploy/helm/kubecf/assets/operations/instance_groups/api.yaml [hunk 1]
```diff
@@ -13,6 +13,14 @@
     docker: docker_app_lifecycle/docker_app_lifecycle.tgz
 {{- end }}
 
+{{if .Values.testing.brain_tests.enabled -}}
+# For brain tests, allow CC access to the insecure registry they will create.
+- type: replace
+  path: /instance_groups/name=api/jobs/name=cloud_controller_ng/properties/cc/diego/insecure_docker_registry_list?
+  value:
+  - "insecure-registry.tcp.((system_domain)):20005"
```
Comment (mook-as) [inline deploy/helm/kubecf/assets/operations/instance_groups/api.yaml]:
```suggestion
  path: /instance_groups/name=api/jobs/name=cloud_controller_ng/properties/cc/diego/insecure_docker_registry_list?/-
  value:
    "insecure-registry.tcp.((system_domain)):20005"
```
This _should_ make it append to any existing lists.  Probably.  Maybe.  Hopefully.

File: deploy/helm/kubecf/assets/operations/instance_groups/diego-cell.yaml [hunk 2]
```diff
@@ -67,6 +67,18 @@
         exec:
           command: [sh, -c, 'ss -nlt sport = 17019 | grep "LISTEN.*:17019"']
 
+{{if .Values.testing.brain_tests.enabled -}}
+# For brain tests, allow garden access to the insecure registry they will create.
+- type: replace
+  path: /instance_groups/name=diego-cell/jobs/name=garden/properties/garden/insecure_docker_registry_list?
+  value:
+  - "insecure-registry.tcp.((system_domain)):20005"
```
Comment (mook-as) [inline deploy/helm/kubecf/assets/operations/instance_groups/diego-cell.yaml]:
```suggestion
  path: /instance_groups/name=diego-cell/jobs/name=garden/properties/garden/insecure_docker_registry_list?/-
  value:
    "insecure-registry.tcp.((system_domain)):20005"
```
Same as above.

File: deploy/helm/kubecf/assets/operations/instance_groups/diego-cell.yaml [hunk 3]
```diff
@@ -67,6 +67,18 @@
...
+- type: replace
+  path: /instance_groups/name=diego-cell/jobs/name=garden/properties/grootfs?/insecure_docker_registry_list
+  value:
+  - "insecure-registry.tcp.((system_domain)):20005"
```
Comment (mook-as) [inline deploy/helm/kubecf/assets/operations/instance_groups/diego-cell.yaml]:
```suggestion
  path: /instance_groups/name=diego-cell/jobs/name=garden/properties/grootfs?/insecure_docker_registry_list/-
  value:
    "insecure-registry.tcp.((system_domain)):20005"
```
One more.

File: deploy/helm/kubecf/templates/brain-tests.yaml [hunk 4]
```diff
@@ -221,4 +221,26 @@ roleRef:
...
+    app.kubernetes.io/version: {{ default .Chart.Version .Chart.AppVersion | quote }}
+    helm.sh/chart: {{ include "kubecf.chart" . }}
+subjects:
+- kind: "ServiceAccount"
+  name: "{{ .Release.Name }}-tests-brain"
+  namespace: {{ .Release.Namespace | quote }}
+roleRef:
+  kind: "ClusterRole"
+  apiGroup: "rbac.authorization.k8s.io"
+  name: "cluster-admin"
```
Comment (mook-as) [inline deploy/helm/kubecf/templates/brain-tests.yaml]: We _should_ be using a less privileged cluster role (just enough to do what we need), but it's brain tests so I don't care _that_ much.

I guess make a note in [`values.yaml`] that enabling tests could potentially be insecure, and they should only be turned on for isolated testing?

[`values.yaml`]: https://github.com/cloudfoundry-incubator/kubecf/blob/4dc4256f0b57c5624f84f0d2222d52fc8390f403/deploy/helm/kubecf/values.yaml#L294

File: deploy/helm/kubecf/assets/operations/instance_groups/api.yaml [hunk 5]
```diff
@@ -13,6 +13,14 @@
     docker: docker_app_lifecycle/docker_app_lifecycle.tgz
 {{- end }}
 
+{{if .Values.testing.brain_tests.enabled -}}
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply deploy/helm/kubecf/assets/operations/instance_groups/api.yaml]: I can switch over and check if that also work. IIRC I had `{{- ... -}}` and that was bad. I then removed the one which seemed to be the cause, which was the leading `-`.

File: deploy/helm/kubecf/assets/operations/instance_groups/api.yaml (hunk 1 above)
Comment (andreas-kupries) [reply deploy/helm/kubecf/assets/operations/instance_groups/api.yaml]: Hm. Can trial. Will be tomorrow, current post-rebase trial is running and EOD when it is done.

File: deploy/helm/kubecf/assets/operations/instance_groups/diego-cell.yaml (hunk 2 above)
Comment (andreas-kupries) [reply deploy/helm/kubecf/assets/operations/instance_groups/diego-cell.yaml]: Ditto.

File: deploy/helm/kubecf/assets/operations/instance_groups/diego-cell.yaml (hunk 3 above)
Comment (andreas-kupries) [reply deploy/helm/kubecf/assets/operations/instance_groups/diego-cell.yaml]: Ditto. Yes, 3 places, all the same thing, with slightly different properties.

File: deploy/helm/kubecf/templates/brain-tests.yaml (hunk 4 above)
Comment (andreas-kupries) [reply deploy/helm/kubecf/templates/brain-tests.yaml]: I am ok with documenting this in values.yaml.
Actually got an ok from Troy when I had the idea to use the big permission hammer here.
I originally tried to simply add permissions until I had all which where needed.
Until the system spat back: `You are trying to grant more permissions than you have`.
That is when I switched to cluster admin to whack this particular rodent.

**Comment by andreas-kupries** [comment]
Ah. Yeah, this PR has to wait for another commit after the BTR PR is merged, tagged, and built before it can be truly merged.

File: deploy/helm/kubecf/assets/operations/instance_groups/api.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply deploy/helm/kubecf/assets/operations/instance_groups/api.yaml]: Addressed with commit  8c3639e66e3f7b076403e5bfcbac35c26769fdd1. Tested to work.

File: deploy/helm/kubecf/templates/brain-tests.yaml (hunk 4 above)
Comment (andreas-kupries) [reply deploy/helm/kubecf/templates/brain-tests.yaml]: Addressed with commit  8c3639e66e3f7b076403e5bfcbac35c26769fdd1. Tested to work. Well, it was just a comment change.

File: deploy/helm/kubecf/templates/brain-tests.yaml [hunk 6]
```diff
@@ -221,4 +221,26 @@ roleRef:
   apiGroup: "rbac.authorization.k8s.io"
   name: "{{ .Release.Namespace }}-{{ .Release.Name }}-test-cluster-role"
 
+# Cluster role binding for service account "tests-brain" and cluster role "cluster-admin"
```
Comment (mook-as) [inline deploy/helm/kubecf/templates/brain-tests.yaml]: We should describe _why_ we need this cluster role binding (as in, the specifics of the tests that require it, not just "for brains 🧟"

Comment (andreas-kupries) [reply deploy/helm/kubecf/templates/brain-tests.yaml]: See commit 8c3639e66e3f7b076403e5bfcbac35c26769fdd1 which has a larger comment in the values.yaml ... Is that one better ? If yes, I can copy the same text over to here.

Comment (mook-as) [reply deploy/helm/kubecf/templates/brain-tests.yaml]: Yes, it is much better :)

Comment (andreas-kupries) [reply deploy/helm/kubecf/templates/brain-tests.yaml]: Ok. Commit copying it coming up

Comment (andreas-kupries) [reply deploy/helm/kubecf/templates/brain-tests.yaml]: Commit aa83f057d613103cdae30cacc3b036784505d591

File: deploy/helm/kubecf/values.yaml [hunk 7]
```diff
@@ -47,7 +47,7 @@ releases:
   bits-service:
     version: 2.28.0
   brain-tests:
-    version: v0.0.8
+    version: v0.0.9
```
Comment (mook-as) [inline deploy/helm/kubecf/values.yaml]: I believe this (for now) needs a stemcell override:
```suggestion
    version: v0.0.9
    stemcell:
      os: SLE_15_SP1
      version: 23.11-7.0.0_374.gb8e8e6af
```

Comment (andreas-kupries) [reply deploy/helm/kubecf/values.yaml]: Ok.

**Comment by andreas-kupries** [comment]
@fargozhu Is the `Status: Do not merge` still relevant ?

**Comment by fargozhu** [comment]
> @fargozhu Is the `Status: Do not merge` still relevant ?

Nope

**Pull request merged**